File: Scripts/Comparison/compare_R_Python_allmodels_600sims.py
"""Cross-language robustness comparison: ToolsRTM (R) vs toolsrtm (Python),
ALL 5 leaf models x ALL 3 canopy models (15 combinations).

Reads each combo's shared LUT + R reflectance written by
compare_R_Python_allmodels_600sims.R, runs the matching Python simulation
for every row, and reports RMSE/R^2/bias/max-abs-diff per combination plus
a combined figure: one simulated-spectra overlay panel per combination
(R vs Python), and a summary heatmap of R^2 across the full leaf x canopy
grid.
"""
import os
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

from toolsrtm import foursail, foursail2, inform

HERE = os.path.dirname(os.path.abspath(__file__))
OUTDIR = os.path.join(HERE, "_out")

LEAF_MODELS = ["PROSPECT-D", "PROSPECT-PRO", "Liberty", "Fluspect-B", "Fluspect-B-Cx"]
CANOPY_MODELS = ["fourSAIL", "foursail2", "INFORM"]

LEAF_EXTRA_COLS = {
    "Liberty": ["cell.d", "inter.c", "baseline.abs", "leaf.thick", "albino.abs", "lign.cell", "Nitrogen"],
    "Fluspect-B": ["Cs", "fqe", "Cx"],
    "Fluspect-B-Cx": ["Cs", "fqe", "Cx"],
}
BASE_COLS = ["N", "Cab", "Car", "Anth", "Cbrown", "EWT", "LMA", "alpha", "Prot", "CBC",
             "LIDFa", "LIDFb", "TypeLidf", "LAI", "hspot", "tts", "tto", "psi"]
CANOPY_EXTRA_COLS = {
    "foursail2": ["fraction_brown", "diss", "Cv", "Zeta"],
    "INFORM": ["LAIu", "sd", "cd", "h", "skyl"],
}


def build_input_row(row, leaf_model, canopy_model):
    cols = list(BASE_COLS) + LEAF_EXTRA_COLS.get(leaf_model, []) + CANOPY_EXTRA_COLS.get(canopy_model, [])
    return {c: row[c] for c in cols}


def run_combo(canopy_model, leaf_model):
    slug = f"{canopy_model.replace('foursail2', 'foursail2') if False else canopy_model}_{leaf_model}"
    slug = "".join(ch for ch in canopy_model if ch.isalnum()) + "_" + "".join(ch for ch in leaf_model if ch.isalnum())
    lut = pd.read_csv(os.path.join(OUTDIR, f"lut_{slug}.csv"))
    refl_R = pd.read_csv(os.path.join(OUTDIR, f"refl_R_{slug}.csv"))
    wl_cols = [c for c in refl_R.columns if c.startswith("wl")]
    wl = np.array([int(c[2:]) for c in wl_cols])
    R_matrix = refl_R[wl_cols].to_numpy()

    is_fluspect = leaf_model in ("Fluspect-B", "Fluspect-B-Cx")
    n = len(lut)
    rsoil_full = np.full(2101, 0.15)
    rsoil_short = np.full(2001, 0.15)
    rsoil_in = rsoil_full if canopy_model == "INFORM" else (rsoil_short if is_fluspect else rsoil_full)

    refl_py = np.full((n, len(wl)), np.nan)
    for i in range(n):
        row = lut.iloc[i]
        inputLUT = build_input_row(row, leaf_model, canopy_model)
        try:
            if canopy_model == "fourSAIL":
                out = foursail(inputLUT, rsoil_in, leaf_model=leaf_model, spectrum_all=not is_fluspect).rsot
            elif canopy_model == "foursail2":
                out = foursail2(inputLUT, rsoil_in, leaf_model=leaf_model, spectrum_all=not is_fluspect).rsot
            else:
                out = inform(inputLUT, rsoil_in, leaf_model=leaf_model)
            if len(out) == len(wl):
                refl_py[i, :] = out
        except Exception:
            pass

    ok = ~np.isnan(refl_py).any(axis=1) & ~np.isnan(R_matrix).any(axis=1)
    R_ok, Py_ok = R_matrix[ok], refl_py[ok]
    diff = Py_ok - R_ok
    n_ok = int(ok.sum())
    if n_ok == 0:
        return dict(canopy_model=canopy_model, leaf_model=leaf_model, n_valid=0,
                     rmse=np.nan, r2=np.nan, bias=np.nan, max_abs=np.nan,
                     wl=wl, R_ok=R_ok, Py_ok=Py_ok)

    rmse = float(np.sqrt(np.mean(diff ** 2)))
    max_abs = float(np.max(np.abs(diff)))
    ss_res = np.sum(diff ** 2)
    ss_tot = np.sum((R_ok - np.mean(R_ok)) ** 2)
    r2 = float(1 - ss_res / ss_tot) if ss_tot > 0 else np.nan
    bias = float(np.mean(diff))
    print(f"{leaf_model:14s} + {canopy_model:9s}: {n_ok}/{n} valid, "
          f"R2={r2:.8f}  RMSE={rmse:.3e}  bias={bias:.3e}  max|diff|={max_abs:.3e}")
    return dict(canopy_model=canopy_model, leaf_model=leaf_model, n_valid=n_ok,
                 rmse=rmse, r2=r2, bias=bias, max_abs=max_abs,
                 wl=wl, R_ok=R_ok, Py_ok=Py_ok)


results = []
for cm in CANOPY_MODELS:
    for lm in LEAF_MODELS:
        results.append(run_combo(cm, lm))

summary = pd.DataFrame([{k: v for k, v in r.items() if k not in ("wl", "R_ok", "Py_ok")} for r in results])
summary.to_csv(os.path.join(OUTDIR, "comparison_allmodels_summary.csv"), index=False)
print("\nWrote comparison_allmodels_summary.csv")
print(summary.to_string(index=False))

# --- Figure 1: R^2 heatmap across the full leaf x canopy grid ---
pivot = summary.pivot(index="canopy_model", columns="leaf_model", values="r2").reindex(
    index=CANOPY_MODELS, columns=LEAF_MODELS)
fig, ax = plt.subplots(figsize=(8, 4))
im = ax.imshow(pivot.values, cmap="RdYlGn", vmin=0.999, vmax=1.0, aspect="auto")
ax.set_xticks(range(len(LEAF_MODELS))); ax.set_xticklabels(LEAF_MODELS, rotation=30, ha="right")
ax.set_yticks(range(len(CANOPY_MODELS))); ax.set_yticklabels(CANOPY_MODELS)
for i in range(len(CANOPY_MODELS)):
    for j in range(len(LEAF_MODELS)):
        val = pivot.values[i, j]
        txt = f"{val:.6f}" if not np.isnan(val) else "n/a"
        ax.text(j, i, txt, ha="center", va="center", fontsize=8)
ax.set_title("R vs Python agreement (R^2) -- every leaf x canopy model combination")
fig.colorbar(im, ax=ax, label="R^2")
plt.tight_layout()
heatmap_path = os.path.join(OUTDIR, "comparison_allmodels_r2_heatmap.png")
plt.savefig(heatmap_path, dpi=150)
plt.close(fig)
print(f"Wrote {heatmap_path}")

# --- Figure 2: simulated-spectra overlay, one panel per combination (R vs Python) ---
fig, axes = plt.subplots(len(CANOPY_MODELS), len(LEAF_MODELS), figsize=(20, 10), sharex=True)
for r in results:
    i = CANOPY_MODELS.index(r["canopy_model"])
    j = LEAF_MODELS.index(r["leaf_model"])
    ax = axes[i, j]
    if r["n_valid"] > 0:
        ax.plot(r["wl"], r["R_ok"][0], color="black", linewidth=1.1, label="R (ToolsRTM)")
        ax.plot(r["wl"], r["Py_ok"][0], color="#B2182B", linewidth=1, linestyle="--", label="Python (toolsrtm)")
        ax.set_title(f"{r['leaf_model']}\n+ {r['canopy_model']}  (R2={r['r2']:.6f})", fontsize=8)
    else:
        ax.set_title(f"{r['leaf_model']}\n+ {r['canopy_model']}  (no valid sims)", fontsize=8)
    if i == len(CANOPY_MODELS) - 1:
        ax.set_xlabel("Wavelength (nm)", fontsize=8)
    if j == 0:
        ax.set_ylabel("TOC reflectance", fontsize=8)
    ax.tick_params(labelsize=7)
handles, labels = axes[0, 0].get_legend_handles_labels()
if handles:
    fig.legend(handles, labels, loc="upper center", ncol=2, bbox_to_anchor=(0.5, 1.02))
fig.suptitle("Simulated reflectance spectra: R (ToolsRTM) vs Python (toolsrtm), all 15 leaf x canopy combinations",
             y=1.05, fontsize=12)
plt.tight_layout()
spectra_path = os.path.join(OUTDIR, "comparison_allmodels_spectra_grid.png")
plt.savefig(spectra_path, dpi=140, bbox_inches="tight")
plt.close(fig)
print(f"Wrote {spectra_path}")

File: Scripts/Comparison/compare_R_Python_allmodels_600sims.R
## Cross-language robustness comparison: ALL leaf models (PROSPECT-D,
## PROSPECT-PRO, Liberty, Fluspect-B, Fluspect-B-Cx) x ALL canopy models
## (fourSAIL, foursail2, INFORM) supported by both ToolsRTM (R) and
## toolsrtm (Python) -- 15 combinations total. Same shared-LUT approach
## as compare_R_Python_toolsrtm_600sims.R, extended to the full matrix
## following the parameter set proven in vignettes/t04-comparing-models.Rmd.

this_file <- sub("^--file=", "", grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE))
root <- normalizePath(file.path(dirname(this_file), "..", ".."))
outdir <- file.path(root, "Scripts/Comparison/_out")
dir.create(outdir, showWarnings = FALSE, recursive = TRUE)

suppressMessages(devtools::load_all(file.path(root, "ToolsRTM"), quiet = TRUE))

n_samples <- 100
wl_2101 <- 400:2500
wl_2001 <- 400:2400

## Structural + PROSPECT-family part of the LUT: real random sampling from
## the package's own parameter-definition table (same source used by the
## single-combination comparison and by the tutorials).
build_base_lut <- function(seed) {
  as.data.frame(ToolsRTM::getLUT(inputs = ToolsRTM::inputsPROSAIL, nLUT = n_samples, setseed = seed))
}

## Extra per-model columns not covered by inputsPROSAIL, jittered around the
## proven-working central values from t04-comparing-models.Rmd's common_lut.
add_liberty_cols <- function(LUT, seed) {
  set.seed(seed)
  n <- nrow(LUT)
  LUT$cell.d        <- runif(n, 25, 80)
  LUT$inter.c       <- runif(n, 0.02, 0.08)
  LUT$baseline.abs  <- runif(n, 0.0004, 0.0008)
  LUT$leaf.thick    <- runif(n, 1.2, 3)
  LUT$albino.abs    <- runif(n, 0, 3)
  LUT$lign.cell     <- runif(n, 1, 6)
  LUT$Nitrogen      <- runif(n, 0.4, 1.8)
  LUT
}
add_fluspect_cols <- function(LUT, seed) {
  set.seed(seed)
  n <- nrow(LUT)
  LUT$Cs  <- runif(n, 0, 0.03)
  LUT$fqe <- runif(n, 0.005, 0.02)
  LUT$Cx  <- runif(n, 0, 0.2)
  LUT
}
add_foursail2_cols <- function(LUT, seed) {
  set.seed(seed)
  n <- nrow(LUT)
  LUT$fraction_brown <- runif(n, 0, 0.3)
  LUT$diss <- runif(n, 0, 1)
  LUT$Cv   <- runif(n, 0.5, 1)
  LUT$Zeta <- runif(n, 0, 0.5)
  LUT
}
add_inform_cols <- function(LUT, seed) {
  set.seed(seed)
  n <- nrow(LUT)
  LUT$LAIu <- runif(n, 0.2, 1.5)
  LUT$sd   <- runif(n, 300, 900)
  LUT$cd   <- runif(n, 2, 7)
  LUT$h    <- runif(n, 10, 30)
  LUT$skyl <- runif(n, 0.05, 0.3)
  LUT
}

leaf_models   <- c("PROSPECT-D", "PROSPECT-PRO", "Liberty", "Fluspect-B", "Fluspect-B-Cx")
canopy_models <- c("fourSAIL", "foursail2", "INFORM")

run_combo <- function(canopy_model, leaf_model, seed) {
  LUT <- build_base_lut(seed)
  if (leaf_model == "Liberty") LUT <- add_liberty_cols(LUT, seed + 1)
  if (leaf_model %in% c("Fluspect-B", "Fluspect-B-Cx")) LUT <- add_fluspect_cols(LUT, seed + 2)
  if (canopy_model == "foursail2") LUT <- add_foursail2_cols(LUT, seed + 3)
  if (canopy_model == "INFORM") LUT <- add_inform_cols(LUT, seed + 4)

  is_fluspect <- leaf_model %in% c("Fluspect-B", "Fluspect-B-Cx")
  ## Output domain: 2001 pts (400-2400 nm) for the two Fluspect leaf models
  ## (their optipar table only spans that range), 2101 pts otherwise --
  ## true for all 3 canopy models alike (inform.R truncates internally,
  ## see r_understorey[1:2001] / r_sail_inf[1:2001]).
  wl <- if (is_fluspect) wl_2001 else wl_2101
  ## rsoil INPUT domain: fourSAIL/foursail2 take rsoil pre-truncated to the
  ## output domain; INFORM always takes the full 2101-length rsoil (its
  ## understorey reflectance is computed at full resolution and only the
  ## final crown-level result gets truncated for Fluspect leaf models).
  rsoil_in <- if (canopy_model == "INFORM") rep(0.15, length(wl_2101)) else rep(0.15, length(wl))

  refl <- matrix(NA_real_, nrow = n_samples, ncol = length(wl))
  for (i in seq_len(n_samples)) {
    row_i <- LUT[i, , drop = FALSE]
    sim <- tryCatch({
      if (canopy_model == "fourSAIL") {
        suppressMessages(foursail(inputLUT = row_i, rsoil = rsoil_in, LeafModel = leaf_model, spectrum.all = !is_fluspect))$rsot
      } else if (canopy_model == "foursail2") {
        suppressMessages(foursail2(inputLUT = row_i, rsoil = rsoil_in, LeafModel = leaf_model))$rsot
      } else {
        suppressMessages(inform(inputLUT = row_i, rsoil = rsoil_in, LeafModel = leaf_model))
      }
    }, error = function(e) NULL)
    if (!is.null(sim) && length(sim) == length(wl)) refl[i, ] <- sim
  }
  ok <- stats::complete.cases(refl)
  cat("R side,", leaf_model, "+", canopy_model, ":", sum(ok), "/", n_samples,
      "valid, domain =", length(wl), "pts\n")
  list(LUT = LUT, refl = refl, ok = ok, wl = wl)
}

summary_rows <- list()
idx <- 0
for (cm in canopy_models) {
  for (lm in leaf_models) {
    idx <- idx + 1
    res <- run_combo(cm, lm, seed = 1000 + idx)
    slug <- paste0(gsub("[^A-Za-z0-9]", "", cm), "_", gsub("[^A-Za-z0-9]", "", lm))
    write.csv(res$LUT, file.path(outdir, paste0("lut_", slug, ".csv")), row.names = FALSE)
    out <- as.data.frame(res$refl); colnames(out) <- paste0("wl", res$wl)
    out$row <- seq_len(n_samples); out$ok <- res$ok
    write.csv(out, file.path(outdir, paste0("refl_R_", slug, ".csv")), row.names = FALSE)
    summary_rows[[idx]] <- data.frame(canopy_model = cm, leaf_model = lm, slug = slug,
                                       n_valid = sum(res$ok), domain_pts = length(res$wl))
  }
}
write.csv(do.call(rbind, summary_rows), file.path(outdir, "allmodels_R_summary.csv"), row.names = FALSE)
cat("Done, all", idx, "leaf x canopy combinations written to", outdir, "\n")
